# Integration and Noise
prior_noise:
  position: 1.0e-3
  rotation: 1.0e-3
  linear_velocity: 1.0e-3
  accelerometer_bias: 1.0e-3
  gyroscope_bias: 1.0e-3
imu_noise:
  integration: 1.0e-3
  integration_bias: 1.0e-3
  accelerometer_bias: 1.0e-3
  gyroscope_bias: 1.0e-3
  overwrite: false # These imu noise values will only be used if overwrite set to true
  accelerometer: 1.0e-3
  gyroscope: 1.0e-3

# Deskew configuration
deskew:
  translation: false
  rotation: false

# ISAM2 Optimisation
isam2:
  optimization: "GaussNewton"  # Options: GaussNewton, Dogleg
  gauss_newton_params:
    wildfire_threshold: 1.0e-3
  dogleg_params:
    initial_delta: 1.0
    wildfire_threshold: 1.0e-5
    trust_region_adaptation_mode: "SEARCH_EACH_ITERATION" # Options: SEARCH_EACH_ITERATION, ONE_STEP_PER_ITERATION
    verbose: false
  relinearize_threshold: 0.1
  relinearize_skip: 10
  enable_relinearization: true
  evaluate_nonlinear_error: false
  factorization: "CHOLESKY" # Options: CHOLESKY, QR
  cache_linearized_factors: true

# Optimisation
optimisation:
  factors:
    imu: true
    registration: true
    stereo: true

# Registration
s2s:
  base_noise:
    rotation: 0.0174533 # 1 deg
    translation: 1.0e-2  # 10 mm
    jacobian_augmentation: true
  method: "fast_gicp"  # Options: icp, gicp, fast_gicp, fast_gicp_st (uses fast_gicp), fast_vgicp, ndt
  base:
    maximum_iterations: 10
    ransac_iterations: 0
    ransac_outlier_rejection_threshold: 0.05
    max_correspondence_distance: 1.0e+9
    transformation_epsilon: 0.0           # Translation convergence threshold in distance units
    transformation_rotation_epsilon: 0.0  # Rotation convergence threshold in cos(angle)
    euclidean_fitness_epsilon: 0.0        # Mean of correspondence euclidean distances convergence threshold 
  icp: # Uses base
    use_reciprocal_correspondences: false
  gicp: # Uses icp
    rotation_epsilon: 2.0e-3
    correspondence_randomness: 20
    maximum_optimizer_iterations: 20
    translation_gradient_tolerance: 1.0e-2  # PCL 1.11.0 onwards
    rotation_gradient_tolerance: 1.0e-2     # PCL 1.11.0 onwards
  lsq: # Uses base
    rotation_epsilon: 2.0e-3
    initial_lambda_factor: 1.0e-9
    debug_print: false
  fast_gicp: # Uses lsq
    num_threads: 4
    correspondence_randomness: 20
    regularization_method: "PLANE" # Options: NONE, MIN_EIG, NORMALIZED_MIN_EIG, PLANE, FROBENIUS
  fast_vgicp: # Uses fast_gicp
    resolution: 1.0
    voxel_accumulation_mode: "ADDITIVE" # Options: ADDITIVE, ADDITIVE_WEIGHTED, MULTIPLICATIVE
    neighbor_search_method: "DIRECT1" # Options: DIRECT27, DIRECT7, DIRECT1, (cuda only) DIRECT_RADIUS
  ndt: # Uses base (note default maximum_iterations for ndt is 35)
    resolution: 1.0
    step_size: 0.1
    outlier_ratio: 0.55

s2m:
  base_noise:
    rotation: 0.0174533 # 1 deg
    translation: 1.0e-2  # 10 mm
    jacobian_augmentation: true
  enabled: true
  method: "fast_gicp"  # Options: icp, gicp, fast_gicp, fast_gicp_st (uses fast_gicp), fast_vgicp, ndt
  base:
    maximum_iterations: 10
    ransac_iterations: 0
    ransac_outlier_rejection_threshold: 0.05
    max_correspondence_distance: 1.0e+9
    transformation_epsilon: 0.0           # Translation convergence threshold in distance units
    transformation_rotation_epsilon: 0.0  # Rotation convergence threshold in cos(angle)
    euclidean_fitness_epsilon: 0.0        # Mean of correspondence euclidean distances convergence threshold 
  icp: # Uses base
    use_reciprocal_correspondences: false
  gicp: # Uses icp
    rotation_epsilon: 2.0e-3
    correspondence_randomness: 20
    maximum_optimizer_iterations: 20
    translation_gradient_tolerance: 1.0e-2  # PCL 1.11.0 onwards
    rotation_gradient_tolerance: 1.0e-2     # PCL 1.11.0 onwards
  lsq: # Uses base
    rotation_epsilon: 2.0e-3
    initial_lambda_factor: 1.0e-9
    debug_print: false
  fast_gicp: # Uses lsq
    num_threads: 4
    correspondence_randomness: 20
    regularization_method: "PLANE" # Options: NONE, MIN_EIG, NORMALIZED_MIN_EIG, PLANE, FROBENIUS
  fast_vgicp: # Uses fast_gicp
    resolution: 1.0
    voxel_accumulation_mode: "ADDITIVE" # Options: ADDITIVE, ADDITIVE_WEIGHTED, MULTIPLICATIVE
    neighbor_search_method: "DIRECT1" # Options: DIRECT27, DIRECT7, DIRECT1, (cuda only) DIRECT_RADIUS
  ndt: # Uses base (note default maximum_iterations for ndt is 35)
    resolution: 1.0
    step_size: 0.1
    outlier_ratio: 0.55

# Mapping
mapping:
  frame_extraction_number: 10
  distance_threshold: 1.0                   # metres
  rotation_threshold: 0.5235987755982988    # radians

# Stereo Factors
stereo_factors:
  noise:
    left_x: 5.0
    right_x: 5.0
    y: 10.0
  left_cam_frame_id: "stereo"
  visualisation:
    print_stats: false
    publish_intermediate_results: false
    rich_keypoints: true
    rich_matches: true
  new_feature_dist_threshold: 5.0
  detector:
    type: "ORB" # Options: GFTT, ORB, SIFT (requires OpenCV >= 4.4.0)
    gftt: &gftt
      max_corners: 1000
      quality_level: 0.01
      min_distance: 1.0
      block_size: 3
      use_harris_detector: false
      k: 0.04
    orb: &orb
      num_features: 500
      scale_factor: 1.2
      num_levels: 8
      edge_threshold: 31
      first_level: 0
      wta_k: 2
      score_type: "HARRIS_SCORE" # Options: HARRIS_SCORE, FAST_SCORE
      patch_size: 31
      fast_threshold: 20
    sift: &sift
      num_features: 10
      num_octave_layers: 3
      contrast_threshold: 0.04
      edge_threshold: 10.0
      sigma: 1.6
  stereo_match_filter:
    vertical_pixel_threshold: 1.0
  sparse_optical_flow:
    type: "PyrLK" # Options: PyrLK (Pyramid Lucas-Kanade), TODO: RLOF
    pyrlk:
      win_size: [21, 21]
      max_level: 3
      term_criteria:
        type: "COUNT+EPS" # Options: COUNT, EPS, COUNT+EPS
        max_count: 30
        epsilon: 0.01
      flags: 0
      min_eig_threshold: 1.0e-4
  stereo_keypoint_matcher:
    cost_function: "SAD" # Options: SAD, SSD, ZMSAD, LSSAD, INCC
    window_size:
      width: 3
      height: 3
    cost_threshold: 1.0e+9
    vertical_pixel_threshold: 1.0
